heroku_connect/models.py

```diff
@@ -0,0 +1,342 @@
+
+    objects = models.Manager.from_queryset(TriggerLogQuerySet)()
+
+    class Meta:
+        abstract = True
+        managed = False
+        get_latest_by = 'created_at'
+        ordering = ('id',)
+
+    is_archived = False
+
+    @classmethod
+    def capture_insert_from_model(cls, table_name, record_id, *, exclude_fields=()):
+        """Create a fresh insert record from the current model state in the database.
+
+        For read-write connected models, this will lead to the attempted creation of a
+        corresponding object in Salesforce.
+
+        Args:
+            table_name: The name of the table backing the connected model (without schema)
+            record_id: The primary id of the connected model
+            exclude_fields: The names of fields that will not be included in the write record
+
+        Raises:
+            LookupError: if ``table_name`` does not belong to a connected model
+
+        """
+        exclude_cols = ()
+        if exclude_fields:
+            model_cls = hc_models.registry.get_class_for_table_name(table_name)
+            exclude_cols = cls._fieldnames_to_colnames(model_cls, exclude_fields)
+
+        sql = """
+            SELECT "{schema}".hc_capture_insert_from_row(
+              hstore({schema}.{table_name}.*),
+              '{table_name}',
+              ARRAY[{exclude_cols}]::text[]
+            ) AS id
+            FROM "{schema}"."{table_name}"
+            WHERE id = %(record_id)s
+        """.format(
+            # TODO escape
+            schema=settings.HEROKU_CONNECT_SCHEMA,
+            table_name=table_name,
+            exclude_cols=', '.join("'{}'".format(col) for col in exclude_cols)
+        )
+        params = {'record_id': record_id}
+        return list(TriggerLog.objects.raw(sql, params))
+
+    @classmethod
+    def capture_update_from_model(cls, table_name, record_id, *, update_fields=()):
+        """Create a fresh update record from the current model state in the database.
+
+        For read-write connected models, this will lead to the attempted update of the values of
+        a corresponding object in Salesforce.
+
+        Args:
+            table_name: The name of the table backing the connected model (without schema)
+            record_id: The primary id of the connected model
+            update_fields: If given, the names of fields that will be included in the write record
+
+        Raises:
+            LookupError: if ``table_name`` does not belong to a connected model
+
+        """
+        include_cols = ()
+        if update_fields:
+            model_cls = hc_models.registry.get_class_for_table_name(table_name)
+            include_cols = cls._fieldnames_to_colnames(model_cls, update_fields)
+        sql = """
+            SELECT "{schema}".hc_capture_update_from_row(
+              hstore({schema}.{table_name}.*),
+              '{table_name}',
+              ARRAY[{include_cols}]::text[]
```

Comment:
type casts can be expensive expensive, is there any way to store these values in the desired types?